standalone/headers-cache/src/main.rs

```diff
@@ -0,0 +1,239 @@
+        #[clap(default_value = "genesis.bin")]
+        output: String,
+    },
+}
+
+#[derive(Subcommand)]
+enum Action {
+    /// Grab genesis or headers from the blockchain and dump it to a file
+    Grab {
+        /// What to grab
+        #[clap(subcommand)]
+        what: Grab,
+    },
+    /// Import genesis or headers from a file into the cache database
+    Import {
+        /// The database file to use
+        #[clap(long, default_value = "cache.db")]
+        db: String,
+        /// What type of data to import
+        #[clap(subcommand)]
+        what: Import,
+    },
+    /// Run the cache server
+    Serve {
+        /// The database file to use
+        #[clap(long, default_value = "cache.db")]
+        db: String,
+    },
+    /// Split given grabbed headers file into chunks
+    Split {
+        /// Size in MB of each chunk
+        #[clap(long, default_value_t = 200)]
+        size: usize,
+
+        /// The headers file to split
+        file: String,
+    },
+    /// Merge given chunks into a single file
+    Merge {
+        /// Appending to existing file
+        #[clap(long, short = 'a')]
+        append: bool,
+        /// The destination file to write to
+        dest_file: String,
+        /// The header chunk files to merge.
+        files: Vec<String>,
+    },
+    /// For debug. Show the authority set id for a given block
+    ShowSetId {
+        #[clap(long)]
+        uri: String,
+        block: BlockNumber,
+    },
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    env_logger::init();
+
+    let args = Args::parse();
+    match args.action {
+        Action::Grab { what } => match what {
+            Grab::Headers {
+                node_uri,
+                para_node_uri,
+                from_block,
+                count,
+                justification_interval,
+                output,
+            } => {
+                let api = pherry::subxt_connect(&node_uri).await?.into();
+                let para_api = pherry::subxt_connect(&para_node_uri).await?.into();
+                let output = File::create(output)?;
+                let count = cache::grap_headers_to_file(
+                    &api,
+                    &para_api,
+                    from_block,
+                    count,
+                    justification_interval,
+                    output,
+                )
+                .await?;
+                println!("{} headers written", count);
+            }
+            Grab::Genesis {
+                node_uri,
+                from_block,
+                output,
+            } => {
+                let api = pherry::subxt_connect(&node_uri).await?.into();
+                let mut output = File::create(output)?;
+                let info = cache::fetch_genesis_info(&api, from_block).await?;
+                output.write_all(info.encode().as_ref())?;
+            }
+        },
+        Action::Import { db, what } => {
+            let mut cache = db::CacheDB::open(&db)?;
```

Comment:
What will happen when importing at the same time when the db is opened for serving? Does leveldb lock the dbfile?

Reply:
I don't think it's necessary. Let's just document the caveat

Reply:
Added in the PR description